docs/design-notes/raft-in-scylla.md

```diff
@@ -0,0 +1,207 @@
+works as a key both internally, when persisting the group state
+or loading it at boot time, and externally, when communicating
+with Raft peers of the group on other nodes.
+
+For example, to persist the changes to schema and topology, Scylla
+can (and does) use node-local system tables.
+There are three such tables:
+    - `raft`, the main table which stores the Raft log,
+    - `raft_snapshots`, a supporting table storing the so-called
+       snapshot descriptors,
+    - `raft_config`, a normalized part of raft
+       `raft_snapshots`, storing the cluster configuration
+       at the time of taking the snapshot. May be out of date with
+       the real cluster configuration, e.g. when configuration
+       change happens after a snapshot and is only present  in
+       `raft` log table.
+
+The system tables are capable of storing the state of an arbitrary
+number of groups, provided the group is happy with the relatively
+low performance the system-table based persistence can provide.
+
+Raft RPC implementation is the same for all groups. It runs on top
+of the standard Scylla messaging service and uses group id as key
+to route messages to the right group on a given host.
+Raft persists peer addresses in its configuration, right next to
+peer identifiers, and the format of the address is not restricted.
+Currently IP addresses are used, just like in the rest of Scylla.
+
+An own client state machine is expected to exit for each group. Right
+now Scylla implements only one client state machine - the so-called
+"schema state machine", created for the only group that contains
+all cluster members - the group 0.
+
+A helper service which can be used to quickly get from a group id
+to its Raft instance is the group registry. It's
+a sharded service which starts early at Scylla boot
+and runs on all shards. The purpose of this component is to
+isolate Raft service consumers within Scylla (i.e. the migration
+manager for schema changes) from the logic of establishing and
+maintaining the group at node start or during topology changes.
+
+# The group 0
+
+The main and the only group running in Scylla now is Group 0. It
+maintains the schema state machine - as stored in the system
+keyspace and the schema cache. Each cluster node is a member of
+this group. The group's Raft server runs on shard 0 on each node.
+Joining and leaving the group is integrated into topology
+operations, and the group is started whenever a node starts.
+
+## Establishing group 0 in a fresh cluster
+
+When a Scylla node starts for the first time, it must join
+group 0. If there is an existing cluster, joining can be done by
+sending a request to add the starting node to the group
+configuration.
+
+But what should the node do if there is no running cluster yet?
+
+A special persistent state machine is responsible for establishing
+an initial Raft configuration in a fresh cluster.
+We often refer to it as "discovery" state machine.
```

Comment:
I think "state machine" is not how one would usually refer to those things. A better term IMO is simply "algorithm".

I would say "A special distributed algorithm is responsible ... We call this the "discovery" algorithm" etc.